darts/explainability/shap_explainer.py

```diff
@@ -0,0 +1,474 @@
+                self.explainers = shap.TreeExplainer(self.model.model)
+
+        self.cache_explainers = None
+        self.cache_foreground_series = None
+        self.cache_foreground_past_covariates = None
+        self.cache_foreground_future_covariates = None
+        self.cache_foreground_X = None
+        self.foreground_changed = True
+
+    def shap_values(
+        self,
+        foreground_series: Union[TimeSeries, Sequence[TimeSeries]],
+        foreground_past_covariates: Union[TimeSeries, Sequence[TimeSeries]],
+        foreground_future_covariates: Union[TimeSeries, Sequence[TimeSeries]],
+        horizon: Optional[int]=None, 
+        target_name: Optional[str]=None
+    ) -> shap._explanation.Explanation:
+        
+        raise_if_not(
+            (len(foreground_series) == len(foreground_past_covariates)) and \
+            (len(foreground_past_covariates) == len(foreground_future_covariates)) and \
+            (len(foreground_future_covariates) == len(foreground_series)),
+            'target, past and future covariates must have the same length.'
+        )
+
+        raise_if_not(
+            all(foreground_series.time_index == foreground_past_covariates.time_index) and \
+            all(foreground_past_covariates.time_index == foreground_future_covariates.time_index) and \
+            all(foreground_future_covariates.time_index == foreground_series.time_index),
+            'The target, past and future covariates must have the same time index.'
+        )
+
+        # We don't recompute if the foreground is the same as the last computation
+        if (self.cache_foreground_series != foreground_series) or \
+        (self.cache_foreground_past_covariates != foreground_past_covariates) or \
+        (self.cache_foreground_future_covariates != foreground_future_covariates):
+
+            foreground_X = self._create_RegressionModel_shap_X(
+                foreground_series,
+                foreground_past_covariates,
+                foreground_future_covariates,
+                None
+            )
+            self.cache_foreground_series = foreground_series
+            self.cache_foreground_past_covariates = foreground_past_covariates
+            self.cache_foreground_future_covariates = foreground_future_covariates
+
+            self.foreground_changed = True
+
+            self.cache_foreground_X = foreground_X.copy()
+        else:
+            self.foreground_changed = False
+
+        return self.shap_values_from_X(
+            self.cache_foreground_X,
+            horizon, 
+            target_name
+            )
+
+    def shap_values_from_X(
+        self, 
+        X,
+        horizon: Optional[int]=None, 
+        target_name: Optional[str]=None
+    ) -> shap._explanation.Explanation:
+        
+        # Multioutput case
+        if (self.target_dim > 1 or self.model.output_chunk_length > 1):
+            
+            # MultioutputRegressor case
+            if isinstance(self.explainers, dict):
```

Comment:
There are many such `isinstance` checks. In general we should try to really keep those at a minimum (they are considered bad practice whenever they can be avoided). Here, I have the feeling that it would be cleaner to make a couple of such checks in the contructor method of the class, once and for all, and then store everything in a unified way. For instance even if the model is made of a single estimator, you can store it as `self.models = {0: estimator}` or something. This way you always have the same format everywhere after that (for instance dict-based), and you don't rely on these checks in other parts of the code.